File: js/chart/marketstructure.js
/* marketstructure.js — BOS and CHoCH.
 *
 * A port of sim/tl/market_structure.py, compared event-for-event in
 * tests/test_ms_parity.py.
 *
 *   BOS    Break of Structure. Price closes through the last swing in the SAME
 *          direction as the prevailing bias — the trend made another leg.
 *
 *   CHoCH  Change of Character. Price closes through the last swing AGAINST the
 *          bias. First evidence the trend may be over, and it flips the bias.
 *
 * The distinction is not in the break — it is the same event — but in the state
 * it arrives in. Closing above the last swing high is a BOS while the bias is
 * already bullish and a CHoCH while it is bearish. That is why this needs a
 * state machine rather than a per-bar rule, and why porting it means porting
 * the STATE, not just the comparison.
 *
 * Definitions follow "Market Structure CHoCH/BOS (Fractal)" by LuxAlgo
 * (TradingView, 2023, open source).
 *
 * CLOSES, NOT WICKS, exactly as tlengine breaks a trendline. And a level is
 * CONSUMED when it breaks: it cannot break twice, so no further bullish event
 * can fire until a new swing high confirms. Without that, one strong trend
 * prints a BOS on every bar that makes a new high.
 */

import { atrSeries } from './tlengine.js';
import { findPivots } from './trendlines.js';

export const BOS = 'bos';
export const CHOCH = 'choch';
export const BULL = 'bullish';
export const BEAR = 'bearish';
export const NEUTRAL = 'neutral';

export const DEFAULT_MS_PARAMS = {
  strength: 3,       // fractal size for the swings
  /* A break must clear the level by this much ATR. 0 is the textbook rule; a
     small value stops a close a tenth of a pip through a level counting as
     structure, which on a 15m chart happens constantly. */
  bufferAtr: 0,
};

/**
 * Returns { events, bias, swingHigh, swingLow, event, eventDir }.
 * Arrays are aligned to bars; `events` is in bar order.
 */
export function detect(bars, params = {}) {
  const p = { ...DEFAULT_MS_PARAMS, ...params };
  const n = bars.length;
  const atr = p.bufferAtr > 0 ? atrSeries(bars, 14) : null;

  const { highs, lows } = findPivots(bars, p.strength);
  /* findPivots carries no confirmedI — it is shared with the batch scorer and
     must stay byte-identical for the other parity tests — so the confirming bar
     is derived here: i + strength by definition. */
  const hiByConf = Array.from({ length: n + 1 }, () => []);
  const loByConf = Array.from({ length: n + 1 }, () => []);
  for (const q of highs) {
    const c = q.i + p.strength;
    if (c >= 0 && c < n) hiByConf[c].push(q);
  }
  for (const q of lows) {
    const c = q.i + p.strength;
    if (c >= 0 && c < n) loByConf[c].push(q);
  }

  let bias = NEUTRAL;
  let sh = null, sl = null;
  const events = [];
  const aBias = new Array(n).fill(NEUTRAL);
  const aSh = new Array(n).fill(NaN);
  const aSl = new Array(n).fill(NaN);
  const aEv = new Array(n).fill('');
  const aDir = new Array(n).fill('');

  for (let i = 0; i < n; i++) {
    for (const q of hiByConf[i]) sh = { price: q.price, i: q.i };
    for (const q of loByConf[i]) sl = { price: q.price, i: q.i };

    let buf = 0;
    if (p.bufferAtr > 0 && atr && Number.isFinite(atr[i])) buf = p.bufferAtr * atr[i];

    const c = bars[i].c;
    if (sh && c > sh.price + buf) {
      const kind = bias === BULL ? BOS : CHOCH;
      const before = bias;
      bias = BULL;
      events.push({ kind, direction: BULL, i, t: bars[i].t, level: sh.price,
                    levelI: sh.i, biasBefore: before, biasAfter: bias, close: c });
      aEv[i] = kind; aDir[i] = BULL;
      sh = null;                       // consumed
    } else if (sl && c < sl.price - buf) {
      const kind = bias === BEAR ? BOS : CHOCH;
      const before = bias;
      bias = BEAR;
      events.push({ kind, direction: BEAR, i, t: bars[i].t, level: sl.price,
                    levelI: sl.i, biasBefore: before, biasAfter: bias, close: c });
      aEv[i] = kind; aDir[i] = BEAR;
      sl = null;
    }

    aBias[i] = bias;
    aSh[i] = sh ? sh.price : NaN;
    aSl[i] = sl ? sl.price : NaN;
  }

  return { events, bias: aBias, swingHigh: aSh, swingLow: aSl,
           event: aEv, eventDir: aDir };
}

/** State at the last bar — what a panel needs. */
export function latest(bars, params = {}) {
  if (!bars || !bars.length) return null;
  const r = detect(bars, params);
  const i = bars.length - 1;
  const last = r.events.length ? r.events[r.events.length - 1] : null;
  return {
    bias: r.bias[i], swingHigh: r.swingHigh[i], swingLow: r.swingLow[i],
    lastEvent: last ? last.kind : null,
    lastEventDir: last ? last.direction : null,
    lastEventI: last ? last.i : null,
    barsSince: last ? i - last.i : null,
  };
}

File: sim/tl/market_structure.py
"""
market_structure.py — BOS and CHoCH.

    BOS    Break of Structure / Market Structure Break.
           Price closes through the last swing in the SAME direction as the
           prevailing bias. Continuation: the trend made another leg.

    CHoCH  Change of Character / Market Structure Shift.
           Price closes through the last swing AGAINST the prevailing bias.
           The first evidence the trend may be over, and it flips the bias.

The distinction is not in the break -- it is the same event -- but in the state
it arrives in. Closing above the last swing high is a BOS while the bias is
already bullish and a CHoCH while it is bearish. That is the whole idea, and it
is why this needs a state machine rather than a per-bar rule.

Definitions follow "Market Structure CHoCH/BOS (Fractal)" by LuxAlgo
(TradingView, 2023, open source): CHoCH confirms a reversal, BOS occurs within
an established trend confirming a new higher high or lower low.

CLOSES, NOT WICKS, exactly as sim/tl/engine.py breaks a trendline: a wick
through the last swing is a test of it, a close through it is a failure of it.

A LEVEL IS CONSUMED WHEN IT BREAKS. Once price closes above the last swing high
that level is spent -- it cannot be broken twice -- and no bullish event can
fire again until a NEW swing high confirms. Without that, one strong trend
prints a BOS on every bar it makes a new high.

CAUSALITY. Only pivots that have reached `confirmed_i` are ever adopted, so the
reference level at bar i is one a bar-i observer could have known. Nothing here
reads forward.
"""

from dataclasses import dataclass
from typing import List, Optional

import numpy as np

from .pivots import find_pivots

BOS = 'bos'
CHOCH = 'choch'
BULL = 'bullish'
BEAR = 'bearish'
NEUTRAL = 'neutral'


@dataclass
class MSEvent:
    """One structural break."""
    kind: str            # bos | choch
    direction: str       # bullish | bearish
    i: int               # bar the close broke through
    t: int               # ms
    level: float         # the swing price that was broken
    level_i: int         # bar the broken swing occurred on
    bias_before: str
    bias_after: str
    close: float

    def to_row(self) -> dict:
        return {'kind': self.kind, 'direction': self.direction, 'i': self.i,
                't': self.t, 'level': self.level, 'level_i': self.level_i,
                'bias_before': self.bias_before, 'bias_after': self.bias_after,
                'close': self.close}


@dataclass
class MSParams:
    strength: int = 3          # fractal size for the swings
    close_confirm: bool = False   # see sim/tl/pivots.py find_pivots
    # A break must clear the level by this much ATR before it counts. Zero
    # reproduces the textbook rule; a small value stops a level being "broken"
    # by a close a tenth of a pip through it, which on a 15m chart happens
    # constantly and produces events nobody would call structure.
    buffer_atr: float = 0.0


def detect(bars, params: MSParams = None, atr=None, times=None):
    """
    Returns (events, arrays).

    `events` is a list of MSEvent in bar order. `arrays` holds per-bar state a
    feature table or a chart can read directly:

        bias        bullish | bearish | neutral, as known at that bar
        swing_high  the live unbroken swing high, NaN once consumed
        swing_low   the live unbroken swing low
        event       'bos' | 'choch' | '' on the bar it fired
        event_dir   'bullish' | 'bearish' | ''
    """
    p = params or MSParams()
    high = np.asarray(bars['high'], dtype=float)
    low = np.asarray(bars['low'], dtype=float)
    close = np.asarray(bars['close'], dtype=float)
    n = len(close)
    if times is None:
        times = np.asarray(bars.index.astype('int64') // 1_000_000)
    if atr is None and p.buffer_atr > 0:
        from ..indicators import atr as atr_series
        atr = atr_series(bars, 14)

    piv_hi, piv_lo = find_pivots(high, low, p.strength, close=close,
                                 close_confirm=getattr(p, 'close_confirm', False))
    # index pivots by the bar they become VISIBLE, not the bar they happened
    hi_by_conf = [[] for _ in range(n + 1)]
    lo_by_conf = [[] for _ in range(n + 1)]
    for q in piv_hi:
        if 0 <= q['confirmed_i'] < n:
            hi_by_conf[q['confirmed_i']].append(q)
    for q in piv_lo:
        if 0 <= q['confirmed_i'] < n:
            lo_by_conf[q['confirmed_i']].append(q)

    bias = NEUTRAL
    sh = sl = None                     # live unbroken swings: {price, i}
    events: List[MSEvent] = []

    a_bias = np.full(n, NEUTRAL, dtype=object)
    a_sh = np.full(n, np.nan)
    a_sl = np.full(n, np.nan)
    a_ev = np.full(n, '', dtype=object)
    a_dir = np.full(n, '', dtype=object)

    for i in range(n):
        # 1. newly visible swings replace the reference level
        for q in hi_by_conf[i]:
            sh = {'price': q['price'], 'i': q['i']}
        for q in lo_by_conf[i]:
            sl = {'price': q['price'], 'i': q['i']}

        buf = 0.0
        if p.buffer_atr > 0 and atr is not None and np.isfinite(atr[i]):
            buf = p.buffer_atr * atr[i]

        # 2. a CLOSE through a live level
        if sh is not None and close[i] > sh['price'] + buf:
            kind = BOS if bias == BULL else CHOCH
            before = bias
            bias = BULL
            events.append(MSEvent(kind, BULL, i, int(times[i]), sh['price'],
                                  sh['i'], before, bias, float(close[i])))
            a_ev[i], a_dir[i] = kind, BULL
            sh = None                  # consumed: it cannot break twice
        elif sl is not None and close[i] < sl['price'] - buf:
            kind = BOS if bias == BEAR else CHOCH
            before = bias
            bias = BEAR
            events.append(MSEvent(kind, BEAR, i, int(times[i]), sl['price'],
                                  sl['i'], before, bias, float(close[i])))
            a_ev[i], a_dir[i] = kind, BEAR
            sl = None

        a_bias[i] = bias
        a_sh[i] = sh['price'] if sh else np.nan
        a_sl[i] = sl['price'] if sl else np.nan

    return events, {'bias': a_bias, 'swing_high': a_sh, 'swing_low': a_sl,
                    'event': a_ev, 'event_dir': a_dir}


def latest(bars, params: MSParams = None):
    """State at the last bar — what a panel needs."""
    events, arr = detect(bars, params)
    if not len(bars):
        return None
    i = len(bars) - 1
    last = events[-1] if events else None
    return {'bias': arr['bias'][i], 'swing_high': arr['swing_high'][i],
            'swing_low': arr['swing_low'][i],
            'last_event': last.kind if last else None,
            'last_event_dir': last.direction if last else None,
            'last_event_i': last.i if last else None,
            'bars_since': (i - last.i) if last else None}
